src/resolvers/dns.rs

```diff
@@ -878,6 +1002,96 @@ mod test {
+        let client = Client::new(
+            &DnsResolverConfig::default(),
+            dns_server.address,
+            Duration::from_millis(100),
+        );
+        match client.lookup_socket_v6(&service).await {
+            Err(err) => match err.kind() {
+                ResolveErrorKind::NoRecordsFound { .. } => (),
+                _ => panic!("unexpected error: {err}"),
+            },
+            Ok(backends) => panic!("unexpectedly found {} backends", backends.len()),
+        }
+
+        dns_server
+            .add_backend(1234, "test001.example.com.", 1000)
+            .await;
+        let backends = client
+            .lookup_socket_v6(&service)
+            .await
+            .expect("successful lookup");
+        assert_eq!(backends.len(), 1);
+
+        dns_server
+            .add_backend(1235, "test002.example.com.", 1000)
+            .await;
+        let backends = client
+            .lookup_socket_v6(&service)
+            .await
+            .expect("successful lookup");
+        assert_eq!(backends.len(), 2);
+
+        dns_server.remove_backend("test001.example.com.").await;
+        let backends = client
+            .lookup_socket_v6(&service)
+            .await
+            .expect("successful lookup");
+        assert_eq!(backends.len(), 1);
+    }
+
+    #[tokio::test]
+    async fn test_no_records_found() {
+        setup_tracing_subscriber();
+
+        // Start a DNS server that reports no records for the service.
+        let mut dns_server = DnsServerBuilder::new("example.com", "test.example.com")
+            .add_backend(1234, "test001.example.com.", 1000)
+            .run_updateable()
+            .await;
+
+        let service = service::Name("test.example.com".into());
+        let bootstrap_servers = vec![dns_server.address];
+        let config = DnsResolverConfig {
+            query_interval: Duration::from_secs(60),
+            query_retry: Duration::from_millis(100),
+            ..Default::default()
+        };
+        let mut resolver = DnsResolver::new(service, bootstrap_servers, config);
+        let mut monitor = resolver.monitor();
+
+        // Wait until we've sen the record, then remove it and wait until our
+        // monitor realizes it's gone (via advancing time past
+        // `query_interval`).
+        wait_for_backends(&mut monitor, 1).await;
+        dns_server.remove_backend("test001.example.com.").await;
+        tokio::time::pause();
+        tokio::time::advance(tokio::time::Duration::from_secs(61)).await;
+        tokio::time::resume();
+        wait_for_backends(&mut monitor, 0).await;
+
+        // Now put the backend back in place on the DNS server side.
+        dns_server
+            .add_backend(1234, "test001.example.com.", 1000)
+            .await;
+
+        // We should nearly immediately find this new backend, due to the very
+        // low `query_retry` interval.
+        wait_for_backends(&mut monitor, 1).await;
```

Comment:
Should we test this with a timeout or something? the comment says "we should nearly immediately" do something, but nothing validates that this happens quickly

Reply:
Yep, easy enough.